\documentclass{bioinfo}
\copyrightyear{2016} \pubyear{2016}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Manuscript Category}

\begin{document}
\firstpage{1}

\subtitle{Subject Section}

\title[short Title]{FUCHS - Full circle characterization RNAseq}
\author[Sample \textit{et~al}.]{Franziska Metge\,$^{\text{\sfb 1,}*}$, Christoph Dieterich\,$^{\text{\sfb 2}*}$}
\address{$^{\text{\sf 1}}$Max-Planck Institute for Biology of Ageing, 50931 Cologne, Germany and \\
$^{\text{\sf 2}}$University Hospital Heidelberg, 69120 Heidelberg, Germany}

\corresp{$^\ast$To whom correspondence should be addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Motivation:} CircRNAs are a type of non-coding RNA neglected by many in the field of next-generation sequencing. Though the amount of circRNA detection tools quickly increased over the past two year, no tool is able to identify fully characterize the circle structure (e.g. exon usage and alternative splicing).\\
\textbf{Results:} Here we present a new method tackling the ful characterization of circRNAs .... and more results about the data.\\
\textbf{Availability:} The method is written as a flexible python pipeline available as git-repository: ....\\
\textbf{Contact:} \href{franziska.metge@age.mpg.de}{franziska.metge@age.mpg.de}\\
\textbf{Supplementary information:}  no supplementary data provided}

\maketitle

\section{Introduction}

\begin{itemize}
\item circle research
\item current programs to detect circles
\item possible functions
\item possible bio-genesis
\item recent RNaseR treatment advances
\item why it would be necessary to know about exon usage and coverage
\end{itemize}

\begin{equation}
\sum \text{\it x}+ \text{\it y} =\text{\it Z}\label{eq:01}\vspace*{-10pt}
\end{equation}
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
\enlargethispage{12pt}

\section{Approach}

\begin{itemize}
\item RNaseR treated cells
\item down stream program after circle detection
\item python based pipeline
\end{itemize}

Equation~(\ref{eq:01}) 
Figure~2\vphantom{\ref{fig:02}} 
\citealp{Boffelli03} 
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\



\begin{methods}
\section{Methods}

.\\\
.\\\
\vspace*{1pt}

\begin{itemize}
\item for bulleted list, use itemize
\item for bulleted list, use itemize
\item for bulleted list, use itemize\vspace*{1pt}
\end{itemize}
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
.\\\
\vadjust{\newpage}.
.\\\
.\\\
.\\\
\subsection{CircRNA detection}
.\\\
.\\\
.\\\
\subsubsection{Mapping}
.\\\
.\\\
.\\\
\subsubsection{CircRNA detection}
.\\\
.\\\
.\\\
\subsection{FUCHS}
% workflow picture
%\subsection{preprocessing}
.\\\
.\\\
.\\\
\subsubsection{Input data}
.\\\
.\\\
.\\\
\subsubsection{Extracting reads}
.\\\
.\\\
.\\\
\subsubsection{Identifying skipped exons}
.\\\
.\\\
.\\\
\subsubsection{Mate pair information, indirect validation of circularity}
% schematic view
.\\\
.\\\
.\\\
\subsubsection{extract coverage}
.\\\
.\\\
.\\\
\subsubsection{Optional steps}
.\\\
.\\\
.\\\
\subsubsection*{isoforms}
.\\\
.\\\
.\\\
\subsubsection*{coverage profile}
.\\\
.\\\
.\\\
\subsubsection*{cluster of circles with similar coverage profile}
.\\\
.\\\
.\\\
\subsection{Visualization}
% elucidate with pictures
.\\\
.\\\
.\\\
\subsubsection{picture from pipeline}
.\\\
.\\\
.\\\
\subsubsection{viusalization in genome browser}
.\\\
.\\\
.\\\
\subsection{Data}
.\\\
.\\\
.\\\
\subsubsection{Mouse heart and liver}
.\\\
.\\\
.\\\
\subsubsection{HEK293 cells}
.\\\
.\\\
.\\\


\begin{table}[!t]
\processtable{This is table caption\label{Tab:01}} {\begin{tabular}{@{}llll@{}}\toprule head1 &
head2 & head3 & head4\\\midrule
row1 & row1 & row1 & row1\\
row2 & row2 & row2 & row2\\
row3 & row3 & row3 & row3\\
row4 & row4 & row4 & row4\\\botrule
\end{tabular}}{This is a footnote}
\end{table}

\end{methods}

\begin{figure}[!tpb]%figure1
\fboxsep=0pt\colorbox{gray}{\begin{minipage}[t]{235pt} \vbox to 100pt{\vfill\hbox to
235pt{\hfill\fontsize{24pt}{24pt}\selectfont FPO\hfill}\vfill}
\end{minipage}}
%\centerline{\includegraphics{fig01.eps}}
\caption{Caption, caption.}\label{fig:01}
\end{figure}

%\begin{figure}[!tpb]%figure2
%%\centerline{\includegraphics{fig02.eps}}
%\caption{Caption, caption.}\label{fig:02}
%\end{figure}



\section{Discussion}
.\\\
.\\\
.\\\









%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%     please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}}
%     as it may ignore the figures.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\section{Results}

.\\\
.\\\
.\\\


\section{Conclusion}
.\\\
.\\\
.\\\

\begin{enumerate}
\item this is item, use enumerate
\item this is item, use enumerate
\item this is item, use enumerate
\end{enumerate}
.\\\
.\\\
.\\\


\section*{Acknowledgements}
.\\\
.\\\
.\\\

\section*{Funding}

.\\\
.\\\
.\\\
\vspace*{-12pt}

%\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}


\begin{thebibliography}{}

\bibitem[Bofelli {\it et~al}., 2000]{Boffelli03}
Bofelli,F., Name2, Name3 (2003) Article title, {\it Journal Name}, {\bf 199}, 133-154.

\bibitem[Bag {\it et~al}., 2001]{Bag01}
Bag,M., Name2, Name3 (2001) Article title, {\it Journal Name}, {\bf 99}, 33-54.

\bibitem[Yoo \textit{et~al}., 2003]{Yoo03}
Yoo,M.S. \textit{et~al}. (2003) Oxidative stress regulated genes
in nigral dopaminergic neurnol cell: correlation with the known
pathology in Parkinson's disease. \textit{Brain Res. Mol. Brain
Res.}, \textbf{110}(Suppl. 1), 76--84.

\bibitem[Lehmann, 1986]{Leh86}
Lehmann,E.L. (1986) Chapter title. \textit{Book Title}. Vol.~1, 2nd edn. Springer-Verlag, New York.

\bibitem[Crenshaw and Jones, 2003]{Cre03}
Crenshaw, B.,III, and Jones, W.B.,Jr (2003) The future of clinical
cancer management: one tumor, one chip. \textit{Bioinformatics},
doi:10.1093/bioinformatics/btn000.

\bibitem[Auhtor \textit{et~al}. (2000)]{Aut00}
Auhtor,A.B. \textit{et~al}. (2000) Chapter title. In Smith, A.C.
(ed.), \textit{Book Title}, 2nd edn. Publisher, Location, Vol. 1, pp.
???--???.

\bibitem[Bardet, 1920]{Bar20}
Bardet, G. (1920) Sur un syndrome d'obesite infantile avec
polydactylie et retinite pigmentaire (contribution a l'etude des
formes cliniques de l'obesite hypophysaire). PhD Thesis, name of
institution, Paris, France.

\end{thebibliography}
\end{document}
